x-pack/plugins/security/server/authentication/authentication_service.ts

```diff
@@ -0,0 +1,207 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import type { PublicMethodsOf } from '@kbn/utility-types';
+import type {
+  LoggerFactory,
+  KibanaRequest,
+  Logger,
+  HttpServiceSetup,
+  IClusterClient,
+  ILegacyClusterClient,
+  HttpServiceStart,
+} from '../../../../../src/core/server';
+import type { SecurityLicense } from '../../common/licensing';
+import type { AuthenticatedUser } from '../../common/model';
+import type { AuditServiceSetup, SecurityAuditLogger } from '../audit';
+import type { ConfigType } from '../config';
+import type { SecurityFeatureUsageServiceStart } from '../feature_usage';
+import type { Session } from '../session_management';
+import type { DeauthenticationResult } from './deauthentication_result';
+import type { AuthenticationResult } from './authentication_result';
+import { getErrorStatusCode } from '../errors';
+import { APIKeys } from './api_keys';
+import { Authenticator, ProviderLoginAttempt } from './authenticator';
+
+interface AuthenticationServiceSetupParams {
+  legacyAuditLogger: SecurityAuditLogger;
+  audit: AuditServiceSetup;
+  getFeatureUsageService: () => SecurityFeatureUsageServiceStart;
+  http: HttpServiceSetup;
+  clusterClient: ILegacyClusterClient;
+  config: ConfigType;
+  license: SecurityLicense;
+  loggers: LoggerFactory;
+  session: PublicMethodsOf<Session>;
+}
+
+interface AuthenticationServiceStartParams {
+  http: HttpServiceStart;
+  clusterClient: IClusterClient;
+}
+
+export interface AuthenticationServiceSetup {
+  /**
+   * @deprecated
+   */
+  getCurrentUser: (request: KibanaRequest) => AuthenticatedUser | null;
+}
+
+export interface AuthenticationServiceStart {
+  apiKeys: Pick<
+    APIKeys,
+    | 'areAPIKeysEnabled'
+    | 'create'
+    | 'invalidate'
+    | 'grantAsInternalUser'
+    | 'invalidateAsInternalUser'
+  >;
+  login: (request: KibanaRequest, attempt: ProviderLoginAttempt) => Promise<AuthenticationResult>;
+  logout: (request: KibanaRequest) => Promise<DeauthenticationResult>;
+  acknowledgeAccessAgreement: (request: KibanaRequest) => Promise<void>;
+  getCurrentUser: (request: KibanaRequest) => AuthenticatedUser | null;
+}
+
+export class AuthenticationService {
+  private license!: SecurityLicense;
+  private authenticator!: Authenticator;
+
+  constructor(private readonly logger: Logger) {}
+
+  setup({
+    legacyAuditLogger: auditLogger,
+    audit,
+    getFeatureUsageService,
+    http,
+    clusterClient,
+    config,
+    license,
+    loggers,
+    session,
+  }: AuthenticationServiceSetupParams): AuthenticationServiceSetup {
+    this.license = license;
+
+    const getCurrentUser = (request: KibanaRequest) => {
+      if (!license.isEnabled()) {
+        return null;
+      }
+
+      return (http.auth.get(request).state ?? null) as AuthenticatedUser | null;
```

Comment:
nit: we can remove the cast:

```suggestion
      return http.auth.get<AuthenticatedUser>(request).state ?? null;
```